### PR TITLE
[Bug Fix] Fix errors in binary cases for roc_auc, precision_recall, and f1 score.

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #541, if available:*

*Description of changes:*
- This PR fixed errors that happen in binary case when using roc_auc, precision_recall_auc, and f1 score, by adding if condition.
- Add unit test cases for evaluation functions.
- Fixed the number of classes in the movielen-100k data, changing it from 6 to 5. Also modified the test scripts according to this number of classes changes.

**Breaking Changes**
This PR unify the behavior of handling errors in evaluation functions, i.e.,  stop code running and raise these errors with corresponding error messages.

Before this PR, some evaluation functions, e.g., roc_auc or f1 score, will return a -1 if some errors happen, but the code can still running. This will give show wrong evaluation results to users. In addition, some errors can still break the running of code, such as roc_auc scores in binary classification. This cause the behavior of handling errors in evaluation different. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/eval/eval_func.py [hunk 1]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
     predicted_labels=logits
     predicted_labels=predicted_labels.detach().cpu().numpy()
     labels=labels.detach().cpu().numpy()
+
+    # check if the two inputs have the same number of rows.
+    try:
+        assert predicted_labels.shape[0] == labels.shape[0]
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:200]: Why not use `assert xxx, "Error message"`?

File: python/graphstorm/eval/eval_func.py [hunk 2]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
+    except AssertionError:
+        logging.error('Predictions and labes should have ' + \
+                f'the same number of records, but got, {predicted_labels.shape[0]} and ' + \
+                f'{labels.shape[0]}.')
+        raise
+
+    # check if the predictions >=2D.
+    try:
+        assert len(predicted_labels.shape) > 1
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:209]: If you assume predicted_labels is a 2D tensor, then `assert len(predicted-lables.shape) == 1`

File: python/graphstorm/eval/eval_func.py [hunk 3]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
+    except AssertionError:
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:210]: ditto

File: python/graphstorm/eval/eval_func.py [hunk 4]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
-        # AUC is only defined when there is at least one positive data.
-        if np.sum(labels[:, i] == 1) > 0 and np.sum(labels[:, i] == 0) > 0:
-            is_labeled = labels[:, i] == labels[:, i]
-            rocauc_list.append(roc_auc_score(labels[is_labeled, i],
-                                             predicted_labels[is_labeled, i]))
+    # For binary results, the sklearn roc_auc_score function asks 1D inputs of predictions.
+    # And the label is a 1D tensor. For other cases, the sklearn roc_auc_score function asks
+    # nD inputs of predictions. So here we need to check the predictions' 2nd dim for the binary
+    # conditions.
+    if len(predicted_labels.shape)>1 and predicted_labels.shape[1] == 2 \
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:222]: We have an assert in L209, we do not need the `len(predicted_labels.shape)>1` check here.

File: python/graphstorm/eval/eval_func.py [hunk 5]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
+        and len(labels.shape) == 1:
+        # Here use the 2nd dim, assuming it is the probability of 1s.
+        rocauc_list.append(roc_auc_score(labels, predicted_labels[:, 1]))
+    else:
+        # mutiple class and multiple labels cases
+        try:
+            labels=labels_to_one_hot(labels, predicted_labels.shape[1])
+        except IndexError as e:
+            logging.error("Failure found during evaluation of the f1 score metric due to" + \
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:231]: f1 score?

File: python/graphstorm/eval/eval_func.py [hunk 6]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
-        if np.sum(labels[:, i] == 1) > 0 and np.sum(labels[:, i] == 0) > 0:
-            is_labeled = labels[:, i] == labels[:, i]
-            rocauc_list.append(roc_auc_score(labels[is_labeled, i],
-                                             predicted_labels[is_labeled, i]))
+    # For binary results, the sklearn roc_auc_score function asks 1D inputs of predictions.
+    # And the label is a 1D tensor. For other cases, the sklearn roc_auc_score function asks
+    # nD inputs of predictions. So here we need to check the predictions' 2nd dim for the binary
+    # conditions.
+    if len(predicted_labels.shape)>1 and predicted_labels.shape[1] == 2 \
+        and len(labels.shape) == 1:
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:223]: What if labels.shape is (x, 1) ?

File: python/graphstorm/eval/eval_func.py [hunk 7]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
+    try:
+        if pred.dim() > 1:
+            # if pred has dimension > 1, it has full logits instead of final prediction
+            assert th.is_floating_point(pred), "Multiple dimension logits are expected to be" + \
+                " float type."
+            pred = pred.argmax(dim=1)
+        # Check if pred is integer tensor
+        assert (not th.is_floating_point(pred) and not th.is_complex(pred)), "1D predictions " + \
+            "are expected to be integer type."
+    except (AssertionError, ValueError):
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:260]: Why we want to capture the asserts?

File: python/graphstorm/eval/eval_func.py [hunk 8]
```diff
@@ -192,62 +194,112 @@ def eval_roc_auc(logits,labels):
...
+        logging.error("Multiple dimension logits are expected to be float type or " + \
+                      "1D predictions are expected to be integer type.")
+        raise
 
+    return th.sum(pred.cpu() == labels.cpu()).item() / len(labels)
 
 def compute_f1_score(y_preds, y_targets):
-    """ compute macro_average f1 score
+    """ compute macro_average f1 score.
+        If any errors occur, log error and return -1, but not interrupt running.
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:269]: Can you add docstring for input arguments and return values?

File: python/graphstorm/eval/eval_func.py [hunk 9]
```diff
@@ -275,12 +327,16 @@ def compute_acc_lp(pos_score, neg_score):
...
 
-
 def compute_roc_auc(y_preds, y_targets, weights=None):
     """ compute ROC's auc score
     """
     y_true = y_targets.cpu().numpy()
     y_pred = y_preds.cpu().numpy()
+
+    # check for binary cases, input 2D and label 1D
+    if len(y_pred.shape) > 1 and y_pred.shape[1] == 2 and len(y_true.shape) == 1:
```
Comment (classicsong) [inline python/graphstorm/eval/eval_func.py:337]: I think roc_auc_score has already handled this case.

File: tests/end2end-tests/data_gen/gen_multilabel.py [hunk 10]
```diff
@@ -35,7 +35,8 @@ def generate_graph_data(data, field):
     new_data = {}
     for name in data:
         if field in name:
-            new_data[name] = gen_multi_label(6, data[name].shape[0])
+            # 01/21/2024 James: correct the number of classes to 5
```
Comment (classicsong) [inline tests/end2end-tests/data_gen/gen_multilabel.py:38]: What is the difference between 5 and 6?

File: tests/end2end-tests/data_gen/process_movielens.py [hunk 11]
```diff
@@ -87,12 +87,14 @@ def write_data_parquet(data, data_file):
 
 # process edges
 edges = pandas.read_csv('/data/ml-100k/u.data', delimiter='\t', header=None)
-edge_data = {'src_id': edges[0], 'dst_id': edges[1], 'rate': edges[2]}
+# 01/21/2024: James change rate to start from 0 to fit evaluation metrics, e.g., roc_auc or p_r
```
Comment (classicsong) [inline tests/end2end-tests/data_gen/process_movielens.py:90]:
```suggestion
# Set the rate to start from 0 to fit evaluation metrics, e.g., roc_auc or p_r
```

File: tests/end2end-tests/data_gen/process_movielens.py [hunk 12]
```diff
@@ -87,12 +87,14 @@ def write_data_parquet(data, data_file):
...
+edge_data = {'src_id': edges[0], 'dst_id': edges[1], 'rate': edges[2]-1}
 write_data_parquet(edge_data, '/data/ml-100k/edges.parquet')
 
 # generate data for homogeneous optimization test
 edges = pandas.read_csv('/data/ml-100k/u.data', delimiter='\t', header=None)
-edge_data = {'src_id': edges[1], 'dst_id': edges[1], 'rate': edges[2]}
+# 01/21/2024: James change rate to start from 0 to fit evaluation metrics, e.g., roc_auc or p_r
```
Comment (classicsong) [inline tests/end2end-tests/data_gen/process_movielens.py:96]: ditto

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 13]
```diff
@@ -24,7 +24,8 @@ error_and_exit () {
 }
 
 echo "**************dataset: Generated multilabel MovieLens EC, RGCN layer: 1, node feat: generated feature, inference: full graph, exclude-training-targets: True"
-python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_multi_label_ec/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_ec.yaml --exclude-training-targets True --multilabel true --num-classes 6 --node-feat-name movie:title user:feat --use-mini-batch-infer false --topk-model-to-save 1  --save-embed-path /data/gsgnn_ec/emb/ --save-model-path /data/gsgnn_ec/ --save-model-frequency 1000 --logging-file /tmp/train_log.txt --logging-level debug --preserve-input True
+# 01/21/2024 James: Change --num-classes to 5
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:27]: Do we need this comment?

File: tests/end2end-tests/graphstorm-ec/test-cpu.sh [hunk 14]
```diff
@@ -33,8 +33,11 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
 error_and_exit $?
 
 # TODO(zhengda) Failure found during evaluation of the auc metric returning -1 multiclass format is not supported
-echo "**************dataset: Test edge classification, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, inference: mini-batch, eval_metric: precision_recall"
-python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ec_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222  --cf ml_ec.yaml --part-config /data/movielen_100k_ec_1p_4t/movie-lens-100k.json --eval-metric precision_recall --num-epochs 1
+# 01/20/2024: (James) change all behavior of evaluation errors as broken, rather than returning -1. So change this
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/test-cpu.sh:36]: Can we just move this comments into PR description?

Comment (classicsong) [reply tests/end2end-tests/graphstorm-ec/test-cpu.sh:36]: It reads like there are some breaking changes in how we handle evaluation errors so we change the tests.

File: python/graphstorm/eval/eval_func.py (hunk 6 above)
Comment (zhjwy9343) [reply python/graphstorm/eval/eval_func.py:223]: if labels.shape is (x, 1), will cause an wrong value. I have changed the code to check if `labels.shape[1] == 1`.

File: python/graphstorm/eval/eval_func.py (hunk 7 above)
Comment (zhjwy9343) [reply python/graphstorm/eval/eval_func.py:260]: The current logics are based on the following situations:
1. As our latest discussion, if anything wrong for the input data of evaluation functions, the code show break and stop.
2. Our test cases should test the invalid cases, including but not limited to wrong dimension, wrong rows, etc.
3. Our test cases should not let the invalid cases break the testing pipeline, i.e., test code should be able to catch errors occurred in the `eval_func.py`.

Based on the three situations, I used the try/catch and raise the assertion errors. Use the current method, these Errors will be raised to function callers. 
- If no one handle these errors, the code will break, just as we expected for users who call our APIs.
- There are test cases in test code to mimic these invalid inputs and can trigger the correct error to happen.
- Our test code catch these errors and handle them properly to let the test pipeline pass through.

File: python/graphstorm/eval/eval_func.py (hunk 9 above)
Comment (zhjwy9343) [reply python/graphstorm/eval/eval_func.py:337]: Sklearn does not handle this case. We need to give the required format for binary cases.

Comment (zhjwy9343) [reply python/graphstorm/eval/eval_func.py:337]: I also modify this condition, including the `y_true.shape` is `(n, 1)` case.

File: python/graphstorm/eval/eval_func.py [hunk 15]
```diff
@@ -359,7 +412,33 @@ def comparator_per_class_roc_auc(best_report, current_report):
...
+        y_preds : Target scores in 2D tensor with shape (n_samples, n_classes).
+        y_targets: the multilple classes case with shape (n_samples, n_classes).
+
+        The number of classes of y_preds should be equal to the number of classes of y_targets.
+        Returns
+        -------
+        A dictionary of auc_roc scores, including average auc_roc score, and score for each class.
+    
     """
+    assert len(y_preds.shape) == 2 and y_preds.shape[1] >= 2, 'ERROR: the given prediction ' + \
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/eval/eval_func.py:428]: These checks are for the for loop in the line 447. The for loop require both preds and targets to be 2D and the 2nd dimension should be >=2 and be the same.

In the line 449, we give roc_auc_score() both 1D arraies.

File: tests/unit-tests/test_eval_func.py [hunk 16]
```diff
@@ -57,15 +57,17 @@ def test_eval_roc_auc():
     labels = th.concat([th.zeros(25), th.ones(25)]).long()
     try:
         error_score_1 = eval_roc_auc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e1:
+        print(f'Test eval_roc_auc error1, {e1}')
         error_score_1 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:63]: Yes, added in the line 115.

File: tests/unit-tests/test_eval_func.py [hunk 17]
```diff
@@ -57,15 +57,17 @@ def test_eval_roc_auc():
...
 
     # Invalid case 2: preds is 1D, return -1 and throws an exception
     preds = th.concat([th.ones(50)*0.25, th.ones(50)*0.75])
     labels = th.concat([th.zeros(50), th.ones(50)]).long()
     try:
         error_score_2 = eval_roc_auc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e2:
+        print(f'Test eval_roc_auc error2, {e2}')
         error_score_2 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:72]: added in the line 116

File: tests/unit-tests/test_eval_func.py [hunk 18]
```diff
@@ -79,13 +81,14 @@ def test_eval_roc_auc():
                         th.ones(25) + 2]).long()
     try:
         error_score_3 = eval_roc_auc(preds, labels)
-    except (AssertionError, ValueError, IndexError):
+    except (AssertionError, ValueError, IndexError) as e3:
+        print(f'Test eval_roc_auc error3, {e3}')
         error_score_3 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:87]: added in the line 117.

File: tests/unit-tests/test_eval_func.py [hunk 19]
```diff
@@ -124,7 +127,8 @@ def test_compute_roc_auc():
     labels = th.concat([th.zeros(25), th.ones(25)]).long()
     try:
         error_score_1 = compute_roc_auc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e1:
+        print(f'Test compute_roc_auc error 1, {e1}')
         error_score_1 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:133]: added in the line 191

File: tests/unit-tests/test_eval_func.py [hunk 20]
```diff
@@ -235,15 +241,17 @@ def test_eval_acc():
...
         
     # Invalid case 2: 1D input in logits format.
     preds = th.concat([th.ones(50) * 0.25, th.ones(50) * 0.75])
     labels = th.concat([th.zeros(50), th.ones(50)]).long()
     try:
         error_acc_2 = eval_acc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e2:
+        print(f'Test eval_acc error 2, {e2}')
         error_acc_2 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:256]: added in the line 192

File: tests/unit-tests/test_eval_func.py [hunk 21]
```diff
@@ -235,15 +241,17 @@ def test_eval_acc():
     labels = th.concat([th.zeros(25), th.ones(25)]).long()
     try:
         error_acc_1 = eval_acc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e1:
+        print(f'Test eval_acc error 1, {e1}')
         error_acc_1 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:247]: added in the line 273.

File: tests/unit-tests/test_eval_func.py [hunk 22]
```diff
@@ -138,7 +142,8 @@ def test_compute_roc_auc():
                         th.ones(25) + 2]).long()
     try:
         error_score_2 = compute_roc_auc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e2:
+        print(f'Test compute_roc_auc error 2, {e2}')
         error_score_2 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:147]: added in the line 193

File: tests/unit-tests/test_eval_func.py [hunk 23]
```diff
@@ -152,7 +157,8 @@ def test_compute_roc_auc():
                         th.ones(25) + 2]).long()
     try:
         error_score_3 = compute_roc_auc(preds, labels)
-    except (AssertionError, ValueError):
+    except (AssertionError, ValueError) as e3:
+        print(f'Test compute_roc_auc error 3, {e3}')
         error_score_3 = -1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:163]: added in the line 193

File: tests/unit-tests/test_eval_func.py [hunk 24]
```diff
@@ -57,15 +57,17 @@ def test_eval_roc_auc():
     labels = th.concat([th.zeros(25), th.ones(25)]).long()
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply tests/unit-tests/test_eval_func.py:57]: Added a test cases in the new line 93

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**